### ISSUE TITLE
Plotting issues/errors with 'plot_allvars' task

### ISSUE THREAD

**Issue by swomantree** [ufs-community/ufs-srweather-app#865]
Hello,
I am an intern working on testing and verification of the UFS-SRW app. Whenever I try to use the COM-OUT_REF  on the plot_all_vars task, I get a weird error.
This is what I'm running in my config.yaml file
task_plot_allvars:
  COMOUT_REF: '{EXPT_BASEDIR}/jun_23_23_12z_rap/${PDY}${cyc}/postprd'

This command should just plot the differences between two experiments. When I use the FV3-HRRR physics, the plot_allvars task fails on hour 36 of my forecast, when it's trying to plot CAPE. 
The log file on my plot_all_vars task gives this error.

ERROR: TopologyException: side location conflict at -101.92278893881431 40.306652768131286
INFO: Self-intersection at or near point -104.51722419282993 42.156323502581415

When the plot_allvars tasks complete on my FV3_RAP run, the plots are not very useful. The plots should have some transparency,  be transparent at 0, or have the map drawn on top of the data for spatial reference. I'm attaching examples of some of the problematic plots.
Thank you for your help.

## Output (optional)
<img width="601" alt="image (3)" src="https://github.com/ufs-community/ufs-srweather-app/assets/106349919/077be64a-385b-426c-9f98-d87b78eb7fac">
<img width="601" alt="image (2)" src="https://github.com/ufs-community/ufs-srweather-app/assets/106349919/c2db16c8-eabc-4cf8-b167-03b9951168e7">
<img width="601" alt="image (4)" src="https://github.com/ufs-community/ufs-srweather-app/assets/106349919/8af83675-72c0-4285-b696-ef2ff0c389c2">

**Comment by gspetro-NOAA** [comment]
Hi @swomantree, 

I am going to see if someone on my team can look into this, but in the meantime, it would be helpful to have a bit more information. Since you're plotting the difference, would it be possible to send us the config files for both experiments? 

Thanks,
Gillian Petro | EPIC Advanced User Support Team

**Comment by gspetro-NOAA** [comment]
Hi @swomantree,

@danielabdi-noaa suggested that the error message could be related to the use of corrupted input data for plotting. The "TopologyException" is a Cartopy error not directly related to the plotting script. If you are able to provide the `config.yaml` script and any other info on the test setup, we can try to reproduce the error, but we do need that information to be able to help you troubleshoot. 

Daniel also thought that the diff plot looked ok (i.e., similar to plots he's gotten in the past), so it seems like perhaps the transparency/overlap over reference area are wishlist items. @BenjaminBlake-NOAA worked on the plotting scripts, so he may be able to provide more information as to whether these look normal and/or whether there is a way to modify the plots for the result you want. 

Best,
Gillian

**Comment by swomantree** [comment]
Hello,
I apologize for the delay in my response. I have been unable to access my
config.yaml file on Hera from my home computer, which is a Windows machine.

 I'm in the process of getting a Mac instead, but as of right now,  I'm not
able to send you that file.

When my mentor gets back from vacation early next week, I can have him send
it to folks.

I apologize for the delay, and as soon as I'm able to access that file,
I'll send it to folks so you can see exactly what's going on.

Thanks for your help,
Sarah Womantree

**Comment by swomantree** [comment]
Hello all,
I was finally able to access those config.yaml files for you folks. I am
attaching them to this email. I have HPSS access, so I'm able to access
most of the available recent data without problems.

I wasn't able to produce any of the difference plots, as the one experiment
failed on hour 36 and there was no data to compare the experiment with to
make the difference plots. Those output plots I uploaded are the default
tasks on the 'COMOUT_REF' part of the config.yaml file. These are the
default 'plot_allvars' outputs of helicity, 10m Temperature, and sea level
pressure. However, there are no terrain features visible and you can't see
the outline of states or any reference locations. They are not helpful, and
I made my own plots using python. I would have liked to produce the
difference plots, just for a quick reference for my research.

I referenced the powerpoint presentation  Ben Blake made  here
<https://dtcenter.org/sites/default/files/events/2021/17-ufs-training-2021-plotting-capabilities.pdf>
using
the UFS plot-allvars script, and his plots look nothing like what is
produced from the default plot_allvars script in the UFS tutorial
documentation. You can see the outline of coasts, states, and terrain
features in his plots, and none of that is visible in the default plots I
uploaded on Github.  Here's an example of his plot
[image: image.png]
and here is a example of my sea level pressure default plot
[image: image.png]

Anyway, I appreciate your help troubleshooting these difference plots. I
ran the default "COMOUT_REF' tasks on the RAP physics, and I ran the
'difference plot' on the 'COMOUT_REF' tasks on the HRRR physics experiment.
Best,
Sarah

**Comment by swomantree** [comment]
Here are the config.yaml files I used to try to create this difference plot.
[config_hrrr.yaml.txt](https://github.com/ufs-community/ufs-srweather-app/files/12372782/config_hrrr.yaml.txt)
[config_rap.yaml.txt](https://github.com/ufs-community/ufs-srweather-app/files/12372784/config_rap.yaml.txt)

**Comment by gspetro-NOAA** [comment]
Thanks for the config files! In general, they look fine, so it might be helpful for me to take a look at the logs. Where are your experiment directories on Hera? Sometimes the final error that causes a failure is actually a result of issues further up in the log, so that might give us an idea of what's going wrong, especially given what Daniel was saying about TopologyException being a Cartopy-specific error.

**Comment by swomantree** [comment]
Yes, my experiment directories are on Hera. I can send you the path for all
of the log files if that's easiest.

**Comment by gspetro-NOAA** [comment]
Yes, that would be helpful! I don't have access to HPSS, so while I'm rectifying that situation, I can take a look at the logs and perhaps get someone with HPSS access to run the experiment/replicate the bug if I'm not able to do so myself.

**Comment by danielabdi-noaa** [comment]
@gspetro-NOAA @swomantree I do have access to HPSS but it looks like tasks are failing at the `get_extrn_ics` step because it could not find the necessary files. The hrrr config file fails with this error message:

```
Username: Daniel.Abdi  UID: 20429  Acct: 20429(20429) Copies: 1 COS: 0 Firewall: off [hsi.9.3.0.p0 Tue Oct 4 14:08:46 UTC 2022] 
*** ls: No such file or directory [-2: HPSS_ENOENT] 
    /NCEPPROD/hpssprod/runhistory/rh2023/202306/20230624/gpfs_hps_nco_ops_com_hrrr_prod_hrrr.20230624_conus12-17.wrf.tar
###WARNING  hsi returned non-zero exit status.
            64 = /apps/hpss/bin/hsi ls /NCEPPROD/hpssprod/runhistory/rh2023/202306/20230624/gpfs_hps_nco_ops_com_hrrr_prod_hrrr.20230624_conus12-17.wrf.tar
WARNING: /NCEPPROD/hpssprod/runhistory/rh2023/202306/20230624/gpfs_hps_nco_ops_com_hrrr_prod_hrrr.20230624_conus12-17.wrf.tar is not available! 
 
INFO: Running command 
 hsi ls /NCEPPROD/hpssprod/runhistory/rh2023/202306/20230624/com_hrrr_prod_hrrr.20230624_conus12-17.wrf.tar 
```
The RAP config file has a typo in the specification of `COMOUT_REF` but the same error occurs after fixing that. Also it is trying to get a "HRRR" ICs file while I expected it to fetch a RAP ICs file. Looking into the HPSS directory, I see files like `com_rap/hrrr_v5..` etc but not `com_rap/hrrr_prod`. Maybe there are changes in HPSS that I am not aware of, or I don't have full access to all files.

**Comment by swomantree** [comment]
Yes, we had to fix some of the paths for the recent HPSS data in order to
run the experiment. My mentor at NOAA/CIRES, Gerard Ketefian, helped me
with all of these fixes. I'm going to forward this conversation to my .noaa
account, and 'cc him.

I can look through our notes of the exact paths we had to amend and where
we changed them in the code. I think he filled out a separate bug report
for the UFS for this problem. We ran into several issues when trying to run
experiments for recent data, but  one of the bug reports/fixes is found here
<https://github.com/ufs-community/ufs-srweather-app/issues/856>,

I think we had to change a path in the data_locations.yml file to fix this
error.
My data_locations.yml file is found here if you want to take a look
(scratch2/BMC/fv3lam/Sarah.Womantree/ufs-srweather-app/parm/data_locations.yml)
but I believe the fix to this error was to add this line:

"the file is parm/data_locations.yml, and you'll have to add this line:

    archive_file_names:
      # HRRR forecasts are binned into 6 hour tar files.
      - gpfs_hps_nco_ops_com_hrrr_prod_hrrr.{yyyymmdd}_conus{bin6}.wrf.tar
      - com_hrrr_prod_hrrr.{yyyymmdd}_conus{bin6}.wrf.tar
     * - com_hrrr_v4.1_hrrr.{yyyymmdd}_conus{bin6}.wrf.tar*

The line in bold is what fixed this error I believe, but Gerard would know
for sure. I thought he made a separate bug report for this particular
problem, so I'll have to check out our notes again. If he didn't, we
probably should make one.

**Comment by swomantree** [comment]
The first time I ran into this error, I ran an experiment starting on June
23, 23. That log file error, which is the same, is found here
  scratch2/BMC/fv3lam/Sarah.Womantree/expt_dirs/june_23_23_12z_hrrr/log
I ran another experiment starting the next day, june_24_23_12z_hrrr, and I
got the same error. I believe that is the config.yaml file I sent you, but
I got the same error regardless.

**Comment by swomantree** [comment]
Hi all,
Sorry to randomly forward a message to you, Gerard. Some of the UFS people
are interested in helping us troubleshoot the plot-allvars failure issue,
but they ran into some of the HPSS issues we ran into when we tried to run
recent experiments. I know you made this bug/issue
<https://github.com/ufs-community/ufs-srweather-app/issues/854> report, but
I can't remember if this fixed the HPSS data issue, or if that was separate.
I went through some of our messages, and the fix to the HPSS data issue was
"the get_extern_ic task is looking for the following file:

/NCEPPROD/hpssprod/runhistory/rh2023/202304/20230419/com_hrrr_prod_hrrr.20230419_conus00-05.wrf.tar

The file on HPSS is the following:

/NCEPPROD/hpssprod/runhistory/rh2023/202304/20230419/com_hrrr_v4.1_hrrr.20230419_conus00-05.wrf.tar

  It's a pain because paths/file names change constantly on HPSS.  To fix
it, you'll need to change this file:
/scratch2/BMC/fv3lam/Robby.Frost/ufs-srweather-app/parm/data_locations.yml

Add the following before line 251:
com_hrrr_v4.1_hrrr.{yyyymmdd}_conus{bin6}.wrf.tar

We had to do something similar for the LBC external model data too on line
225 of data_locations.yml
      - ./
    archive_file_names:
      # RAP forecasts are binned into 6 hour tar files.
      - gpfs_hps_nco_ops_com_rap_prod_rap.{yyyymmdd}{bin6}.wrf.tar
      - com_rap_prod_rap.{yyyymmdd}{bin6}.wrf.tar
     * - com_rap_v5.1_rap.{yyyymmdd}{bin6}.wrf.tar*
    file_names: &rap_file_names

The fix to this error was to move  *-
com_rap_v5.1_rap.{yyyymmdd}{bin6}.wrf.tar *to the top of the list in
data_locations.yml

Maybe you can provide more clarity or insight into what we had to do to fix
this HPSS issue so they can troubleshoot my plot_allvars issue.
Best,
Sarah

**Comment by gsketefian** [comment]
@danielabdi-noaa @gspetro-NOAA @swomantree During Sarah's internship this summer, we encountered (at least) a couple of issues with the `get_extr_[ics|lbcs]` tasks.  I documented them at #854 and #856.

@danielabdi-noaa If you haven't resolved the above problem yet, I'm happy to help if you send me your experiment and clone paths (hopefully you're working on Hera or Jet).  Thanks.

**Comment by danielabdi-noaa** [comment]
Thank you @gsketefian @swomantree I copied the `data_locations.yaml` file from Sarah's directory that had the changes you made and it worked for me after that. Yes, this seems to be a  critical bug that needs to be fixed soon.
Anway, I am running the HRRR test case and comparing against Sarah's RAP run and will let you know how it goes later.
I don't think the problem lies with the plotting scripts but with the output data used plotting, mainly because the non-diff plots worked for Sarah for all 48 hrs for the RAP run, and upto 36 hrs for the HRRR run.

**Comment by danielabdi-noaa** [comment]
@swomantree The HRRR run seems to work for me. Here is my experiment directory:
```
/scratch2/BMC/gsd-hpcs/Daniel.Abdi/expt_dirs/june_24_23_12z_hrrr
```
The plot task run out of wallclock time but it was going well until then. I may try to re-run it with a larger wallclock time limit for the plot task to see if it can produce all diff plots. Despite what I said earlier, your HRRR run did not actually start making diff plots, it failed at the 36th hour of the regular plot task. Also your `COMOUT_REF` path should not use `EXPTDIR` since the experiment directory we want to compare to is that of RAP's, so set it explicitly like this:
```
COMOUT_REF: '/scratch2/BMC/fv3lam/Sarah.Womantree/expt_dirs/june_24_23_12z_rap/${PDY}${cyc}/postprd'
```

**Comment by danielabdi-noaa** [comment]
After increasing walltime for `plot_allvars` task to 08:00:00 hrs, both the regular and diff plots files finished in about 2 hrs.
I increased the walltime directly in the worklflow xml file for this run, but you would need to add something like this in the config file under `rocoto: tasks:` section.
```
task_plot_allvars:
   walltime: 08:00:00
```
The plots are here:
```
/scratch2/BMC/gsd-hpcs/Daniel.Abdi/expt_dirs/june_24_23_12z_hrrr/2023062412/postprd
```

**Comment by swomantree** [comment]
Nice, thank you Daniel! I'll try that out. I'm trying to run a new
experiment, so I'll see if the SRW app can produce those regular and
difference plots now.
